### PR TITLE
Add support for loggerFunction

### PR DESCRIPTION
**Pull request by Slava**
Allows graphql to log whenever the executor enters or leaves a subtree, as well
as when an error occurs.

### CODE REVIEW

**Comment by Slava** [comment]
@helfer can you please code-review it?

**Comment by helfer** [comment]
I just had another thought: I think it would be great if we could make `subtree.start`, `subtree.finish` etc.  string constants that graphql-js exports. That way logger functions can filter events with `===`. That would also mean that instrumenting `resolver.start` and `resolver.end` doesn't cost us anything - i.e. I'm arguing for adding those.

**Comment by Slava** [comment]
@helfer can you please expand on the `resolver.start` and `resolver.end`? How are those relevant to the exported enums?

**Comment by helfer** [comment]
I was just trying to say that we should add instrumentation for those events as well, because it won't slow down the code at all, if loggerFunctions can have a simple equality check to decide whether or not they care about a certain event being logged. If they don't care about resolver events, they can just filter them out.

**Comment by Slava** [comment]
@helfer updated the PR, can you please review again?

File: src/graphql.js [hunk 1]
```diff
@@ -39,14 +39,19 @@ import type { GraphQLSchema } from './type/schema';
...
  */
 export function graphql(
   schema: GraphQLSchema,
   requestString: string,
   rootValue?: mixed,
   contextValue?: mixed,
   variableValues?: ?{[key: string]: mixed},
-  operationName?: ?string
+  operationName?: ?string,
+  loggerFunction?: (tag: string, payload: mixed, info: mixed) => void
```
Comment (helfer) [inline src/graphql.js]: I would just call this `logFunction`

**Comment by helfer** [comment]
I think this does exactly what we need. Let's make the PR to see what Lee Byron says.

One thing that might be useful is to add an event that gets logged from inside the resolve function, just so we know it's working and so we have a code example to show people. Having the subtree and resolver events is the icing on the cake for us, but I think the main thing that's going to convince people is the ability to log from inside the resolvers, so we should have a clear example of that and describe that first in the PR, before we describe the other parts (even though those other parts are 90% of the PR).

**Comment by Slava** [comment]
I think the logs from a resolver should be done by something external like Tracer-Agent. If we keep this PR's scope to "this only logs the events the user of graphql module might care about", I think we are set and done here. I will squash and open a PR.

**Pull request merged**